### PR TITLE
Feature: adding grpc to receive mongo order status histories

### PR DESCRIPTION
**Pull request by eapg**
* Adding Grpc Service to insert mongo Order Status histories from pyrhon Etl
* Addong mongo Order Status History Service to insert mongo order status histories into MongoDB

### CODE REVIEW

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.ea.restaurant.constants.EtlStatus;
+import com.ea.restaurant.constants.OrderStatus;
+import com.ea.restaurant.constants.Status;
+import com.ea.restaurant.document.MongoOrderStatusHistory;
+import java.util.ArrayList;
+import java.util.List;
+import org.bson.types.ObjectId;
+
+public class GrpcMongoOrderStatusHistoryMapper {
+  public static List<MongoOrderStatusHistory> mapGrpcMongoOrderStatusToMongoOrderStatusHistoryList(
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java:12]: Done.

File: src/main/java/com/ea/restaurant/util/TimeUtil.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.ea.restaurant.util;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+public class TimeUtil {
+  public static Instant convertMillisecondsToInstant(Long milliseconds) {
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/util/TimeUtil.java:7]: Done.

File: src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  @Override
+  public void insertMongoOrderStatusHistoriesFromPythonEtl(
+      MongoOrderStatusHistoriesFromPythonRequest request,
+      StreamObserver<NotResponse> responseObserver) {
+    var mappedMongoOrderStatusHistories =
+        GrpcMongoOrderStatusHistoryMapper.mapGrpcMongoOrderStatusToMongoOrderStatusHistoryList(
+            request.getMongoOrderStatusHistoryList());
+    this.mongoOrderStatusHistoryService.insertMongoOrderStatusHistories(
+        mappedMongoOrderStatusHistories);
+    responseObserver.onNext(NotResponse.newBuilder().build());
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java:28]: Return the new mongo ids to the client

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        .build();
+  }
+
+  public static List<MongoOrderStatusHistory> mapGrpcMongoOrderStatusToMongoOrderStatusHistoryList(
+      List<com.ea.restaurant.grpc.MongoOrderStatusHistory> grpcMongoOrderStatusHistories) {
+    var mongoOrderStatusHistories = new ArrayList<MongoOrderStatusHistory>();
+    for (com.ea.restaurant.grpc.MongoOrderStatusHistory mongoOrderStatusHistory :
+        grpcMongoOrderStatusHistories) {
+      mongoOrderStatusHistories.add(
+          mapGrpcMongoOrderStatusToMongoOrderStatusHistory(mongoOrderStatusHistory));
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java:38]: No need if this if ypu use streams

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.ea.restaurant.grpc.MongoOrderStatusHistoryServiceGrpc.MongoOrderStatusHistoryServiceImplBase;
+import com.ea.restaurant.service.impl.MongoOrderStatusHistoryServiceImpl;
+import com.ea.restaurant.util.GrpcMongoOrderStatusHistoryMapper;
+import io.grpc.stub.StreamObserver;
+import net.devh.boot.grpc.server.service.GrpcService;
+
+@GrpcService
+public class GrpcMongoOrderStatusHistoryServiceImpl extends MongoOrderStatusHistoryServiceImplBase {
+
+  private final MongoOrderStatusHistoryServiceImpl mongoOrderStatusHistoryService;
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java:12]: Why do you point to the impl instead of using the interface? Remember the principles of the DI. Do not use the implementation, use the interface or abstract class instead

File: src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java [hunk 6]
```diff
@@ -32,4 +32,10 @@ public void updateBatchToProcessed(Set<ObjectId> ids) {
     }
     this.mongoOrderStatusHistoryRepository.saveAll(mongoOrderStatusHistories);
   }
+
+  @Override
+  public void insertMongoOrderStatusHistories(
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java:37]: I think this should be transactional including a rollback for.

Also this should be made in another pr not in this one

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        .updatedBy(mongoOrderStatusHistory.getUpdatedBy())
+        .createdDate(Instant.ofEpochMilli(mongoOrderStatusHistory.getCreatedDate()))
+        .updatedDate(Instant.ofEpochMilli(mongoOrderStatusHistory.getUpdatedDate()))
+        .build();
+  }
+
+  public static List<MongoOrderStatusHistory> mapGrpcMongoOrderStatusToMongoOrderStatusHistoryList(
+      List<com.ea.restaurant.grpc.MongoOrderStatusHistory> grpcMongoOrderStatusHistories) {
+    var mongoOrderStatusHistories = new ArrayList<MongoOrderStatusHistory>();
+    for (com.ea.restaurant.grpc.MongoOrderStatusHistory mongoOrderStatusHistory :
```
[the remark this answers is not in the document]
Comment (eapg) [reply src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java:35]: Done.

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java (hunk 4 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java:38]: Done. It was Replace.

File: src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java (hunk 6 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/service/impl/MongoOrderStatusHistoryServiceImpl.java:37]: Done.

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        .createdDate(Instant.ofEpochMilli(mongoOrderStatusHistory.getCreatedDate()))
+        .updatedDate(Instant.ofEpochMilli(mongoOrderStatusHistory.getUpdatedDate()))
+        .build();
+  }
+
+  public static List<MongoOrderStatusHistory> mapGrpcMongoOrderStatusToMongoOrderStatusHistoryList(
+      List<com.ea.restaurant.grpc.MongoOrderStatusHistory> grpcMongoOrderStatusHistories) {
+    return grpcMongoOrderStatusHistories.stream()
+        .map(GrpcMongoOrderStatusHistoryMapper::mapGrpcMongoOrderStatusToMongoOrderStatusHistory)
+        .collect(Collectors.toList());
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryMapper.java:34]: I think you can use `toList()` instead of `collect()` here.

File: src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryUtil.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class GrpcMongoOrderStatusHistoryUtil {
+
+  public static List<String> getMongoOrderStatusUuids(
+      List<MongoOrderStatusHistory> mongoOrderStatusHistoryList) {
+    return mongoOrderStatusHistoryList.stream()
+        .map(mongoOrderStatusHistory -> mongoOrderStatusHistory.getId().toString())
+        .collect(Collectors.toList());
```
Comment (reynoldmorel) [inline src/main/java/com/ea/restaurant/util/GrpcMongoOrderStatusHistoryUtil.java:13]: I think you can use toList() instead of collect() here.

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java (hunk 3 above)
Comment (eapg) [reply src/main/java/com/ea/restaurant/grpc/GrpcMongoOrderStatusHistoryServiceImpl.java]: Done.

**Review by reynoldmorel — APPROVED** [review]
